### ISSUE TITLE
Metal: atomics intrinsics.

### ISSUE THREAD

**Issue by csyonghe** [shader-slang/slang#4026]

**Comment by jkwak-work** [comment]
I learned following facts,
The atomic operations works only with atomic data types such as,
 - atomic_int
 - atomic_uint
 - atomic_bool
 - atomic_ulong
 - atomic_float

The assignment to the atomic data types need to happen with an explicit store function
 - atomic_store_explicit

One thing to note is that there is an alternative way to describe the atomic types in a form of `atomic<int>` but it is available only at 2.4 or above. Our current Metal profile is 2.3 and the alternative form is not available.

**Comment by csyonghe** [comment]
For buffers, it looks like we can just emit the c like cast to cast a pointer to an atomic pointer type.

**Comment by jkwak-work** [comment]
I like to update my progress.
I am currently trying to promote the type from the basic type to its atomic type if they are used in the atomic functions.

1. added new IRInst-s.
```
    INST(MetalAtomicAdd2, MetalAtomicAdd2, 2, 0)
    INST(MetalAtomicAdd3, MetalAtomicAdd3, 3, 0)
```
2.  Call them from the existing InterlockedXXX function.
```
__intrinsic_op($(kIROp_MetalAtomicAdd2))
void InterlockedAdd_metal(__ref uint dest, uint value);

void InterlockedAdd(__ref uint dest, uint value)
{
    __target_switch
    {
    case hlsl: __intrinsic_asm "InterlockedAdd";
    case metal:
        InterlockedAdd_metal(dest, value);    // <== HERE
        return;
```

3. Iterate all IRInst in "slang-ir-metal-legalize.cpp", find "kIROp_MetalAtomicAdd2" and trace back where its first argument came from.
4. Replace the traced variables to its atomic version.

I haven't gotten to the buffers yet.

**Comment by csyonghe** [comment]
I think you can define 

```
__intrinsic_op(kIROp_MetalAtomicCast)
Ref<T> __getMetalAtomicRef(__ref T x);
```

to conceptually convert a normal pointer into a corresponding atomic type pointer.

Then InterlockedAdd(x, v) can just call:
```
__metalInterlockedAdd(__getMetalAtomicRef(x), v);

```

Where `metalInterlockedAdd` simply maps to `intrinisc_asm "(*$0).interlocked_add(v)"`.

**Comment by jkwak-work** [comment]
Oh. That's a great idea. I will do that

**Comment by jkwak-work** [comment]
> Make sure PtrCast or bit cast is emitted as a normal c style cast in the metal emit logic. Metal does not allow static_cast/reinterpret_cast on pointer types, but c like casting works.

I just tried this and it works.
It seems little odd that the underlying data type doesn't actually need to be an atomic type but I will go with it.

**Comment by jkwak-work** [comment]
I am having a trouble with the type `Ref<T>`.
When I tried the following, I get a compile error,
```
__generic<T>
__intrinsic_op($(kIROp_MetalAtomicCast))
Ref<T> __getMetalAtomicRef(__ref T x);

__generic<T>
void __metalInterlockedAdd(Ref<T> dest, uint value)
{
    __intrinsic_asm "atomic_fetch_add_explicit($0, $1, memory_order_relaxe)";
}
```
The error message indicated that `__metalInterlockedAdd()` received "(uint, uint)" not "(Ref<uint>,uint)" when I used it as following,
```
    case metal:
        __metalInterlockedAdd(__getMetalAtomicRef(dest), value);
        return;
```
It seems like the result from `__getMetalAtoicRef` is stored in a temporary variable whose type is just `T` not `Ref<T>`.
And it is sent to `__metalInterlockedAdd`.

As an alternative, I also tried the following change and it does compile to metallib.
```
__generic<T>
void __metalInterlockedAdd(__ref T dest, T value)
{
    __intrinsic_asm "atomic_fetch_add_explicit(((threadgroup atomic_$[0]*)($0)), $1, memory_order_relaxed)", T;
}
```
The problem with this alternative is that the keyword `threadgroup` is hardcoded.
It appears that there are overloading for the atomic functions for `threadgroup` and `device`.
So this alternative is not good enough.
I will need to figure out why `Ref<T>` is not working as expected.

**Comment by csyonghe** [comment]
I should be fine if you just use 

T* getAtomicPtr(__ref T x);

**Comment by csyonghe** [comment]
And you can do

metalInterlockedAdd(T* ptr, U v)
{
     intrinsic_asm "atomic_add($0, $1)";
}

T should already have an address space when you reach the emit logic for this function, because the argument passes to ptr should be a kIROp_MetalAtomicCast, and we should be propagating the address space through these insts during legalization.

**Comment by jkwak-work** [comment]
PR is merged as #4473.
Closing the issue
